Description: fix invalid reads when the class name is too short
Origin: https://github.com/phalcon/cphalcon/pull/849
Bug: https://github.com/phalcon/cphalcon/pull/849
Forwarded: https://github.com/phalcon/cphalcon/pull/849
Last-Update: 2012-06-16

--- php-phalcon-1.2.0.orig/cphalcon/ext/kernel/alternative/fcall.c
+++ php-phalcon-1.2.0/cphalcon/ext/kernel/alternative/fcall.c
@@ -915,7 +915,7 @@ int phalcon_alt_call_user_method(zend_cl
 		}
 
 		/** Create a unique key */
-		if (ce->name[7] == '\\') {
+		if (ce_name_length >= 8 && ce->name[7] == '\\') {
 
 			for (i = 7; i < ce->name_length; i++) {
 				hash_key = ce->name[i] + (hash_key << 6) + (hash_key << 16) - hash_key;
--- php-phalcon-1.2.0.orig/cphalcon/build/32bits/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/32bits/phalcon.c
@@ -11780,7 +11780,7 @@ static int phalcon_alt_call_user_method(zend_class_entry *ce, zval **object_pp,
 		}
 
 		/** Create a unique key */
-		if (ce->name[7] == '\\') {
+		if (ce->name_length >= 8 && ce->name[7] == '\\') {
 
 			for (i = 7; i < ce->name_length; i++) {
 				hash_key = ce->name[i] + (hash_key << 6) + (hash_key << 16) - hash_key;
--- php-phalcon-1.2.0.orig/cphalcon/build/64bits/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/64bits/phalcon.c
@@ -11780,7 +11780,7 @@ static int phalcon_alt_call_user_method(zend_class_entry *ce, zval **object_pp,
 		}
 
 		/** Create a unique key */
-		if (ce->name[7] == '\\') {
+		if (ce->name_length >= 8 && ce->name[7] == '\\') {
 
 			for (i = 7; i < ce->name_length; i++) {
 				hash_key = ce->name[i] + (hash_key << 6) + (hash_key << 16) - hash_key;
--- php-phalcon-1.2.0.orig/cphalcon/build/safe/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/safe/phalcon.c
@@ -11780,7 +11780,7 @@ static int phalcon_alt_call_user_method(zend_class_entry *ce, zval **object_pp,
 		}
 
 		/** Create a unique key */
-		if (ce->name[7] == '\\') {
+		if (ce->name_length >= 8 && ce->name[7] == '\\') {
 
 			for (i = 7; i < ce->name_length; i++) {
 				hash_key = ce->name[i] + (hash_key << 6) + (hash_key << 16) - hash_key;
